octopusdeploy/accounts.go

```diff
@@ -0,0 +1,139 @@
+package octopusdeploy
+
+import (
+	"fmt"
+	"github.com/OctopusDeploy/go-octopusdeploy/octopusdeploy"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func getCommonAccountsSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"name": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"description": {
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"environments": {
+			Type: schema.TypeList,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+			Optional: true,
+		},
+		"tenanted_deployment_participation": getTenantedDeploymentSchema(),
+		"tenants": {
+			Type: schema.TypeList,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+			Optional: true,
+		},
+		"tenant_tags": {
+			Type: schema.TypeList,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+			Optional: true,
+		},
+	}
+}
+
+func fetchAndReadAccount(d *schema.ResourceData, m interface{}) (*octopusdeploy.Account, error) {
+	client := m.(*octopusdeploy.Client)
+
+	accountID := d.Id()
+	account, err := client.Account.Get(accountID)
+
+	if err == octopusdeploy.ErrItemNotFound {
+		d.SetId("")
+		return nil, fmt.Errorf("account %s not found: %s ", accountID, err.Error())
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error readingaccount %s: %s", accountID, err.Error())
+	}
+
+	d.Set("name", account.Name)
+	d.Set("description", account.Description)
+	d.Set("environments", account.EnvironmentIDs)
+	d.Set("tenanted_deployment_participation", account.TenantedDeploymentParticipation)
+	// TODO TenantID should be named TenantIds and be a string array
+	d.Set("tenants", account.TenantID)
+	d.Set("tenant_tags", account.EnvironmentIDs)
+
+	return account, nil
+}
+
+func buildAccountResourceCommon(d *schema.ResourceData, accountType octopusdeploy.AccountType) *octopusdeploy.Account {
+	var account = octopusdeploy.NewAccount(d.Get("name").(string), accountType);
+
+
+	if v, ok := d.GetOk("tenant_tags"); ok {
+		account.TenantTags = getSliceFromTerraformTypeList(v)
+	}
+
+	if v, ok := d.GetOk("tenanted_deployment_participation"); ok {
+		account.TenantedDeploymentParticipation, _ = octopusdeploy.ParseTenantedDeploymentMode(v.(string))
+	}
+
+	// TODO TenantID should be named TenantIds and be a string array
+	//if v, ok := d.GetOk("tenants"); ok {
+	//	//account.TenantID = getSliceFromTerraformTypeList(v)
+	//}
+
+	if v, ok := d.GetOk("tenant_tags"); ok {
+		account.TenantTags = getSliceFromTerraformTypeList(v)
+	}
+
+
+	return account;
+}
+
+func resourceAccountCreateCommon(d *schema.ResourceData, m interface{}, account *octopusdeploy.Account) error {
```

Comment:
I don't know go well enough, to implement this without research. The Create/Update/Delete functions all do the same things except for the one line. In C# I would create a common method and pass
```
client => client.Account.Add(account)
```